contracts/multisigs/SafeMultisigWithRecoveryModule.sol

```diff
@@ -0,0 +1,107 @@
+interface ISafeProxyFactory {
+    /// @dev Allows to create new proxy contact and execute a message call to the new proxy within one transaction.
+    /// @param _singleton Address of singleton contract.
+    /// @param initializer Payload for message call sent to new proxy contract.
+    /// @param saltNonce Nonce that will be used to generate the salt to calculate the address of the new proxy contract.
+    function createProxyWithNonce(
+        address _singleton,
+        bytes memory initializer,
+        uint256 saltNonce
+    ) external returns (address proxy);
+}
+
+/// @dev Provided zero address.
+error ZeroAddress();
+
+/// @dev Provided incorrect data length.
+/// @param expected Expected minimum data length.
+/// @param provided Provided data length.
+error IncorrectDataLength(uint256 expected, uint256 provided);
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @title SafeMultisigWithRecoveryModule - Smart contract for Safe multisig creation with the recovery module
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract SafeMultisigWithRecoveryModule {
+    // Selector of the Safe setup function
+    bytes4 public constant SAFE_SETUP_SELECTOR = 0xb63e800d;
+    // Encoded selector of the Recovery module enableModule function
+    bytes4 public constant ENABLE_MODULE_SELECTOR = 0x24292962;
+    // Default data size for several Safe Factory params without payload: address + uint256 = 2 full slots = 64 (bytes)
+    uint256 public constant DEFAULT_DATA_LENGTH = 64;
+
+    // Safe contract address
+    address public immutable safe;
+    // Safe Factory contract address
+    address public immutable safeProxyFactory;
+    // Recovery module address
+    address public immutable recoveryModule;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    /// @dev SafeMultisigWithRecoveryModule constructor.
+    /// @param _safe Safe contract address.
+    /// @param _safeProxyFactory Safe proxy factory contract address.
+    /// @param _recoveryModule Recovery module address.
+    constructor (address _safe, address _safeProxyFactory, address _recoveryModule) {
+        // Check for zero addresses
+        if (_safe == address(0) || _safeProxyFactory == address(0) || _recoveryModule == address(0)) {
+            revert ZeroAddress();
+        }
+
+        safe = _safe;
+        safeProxyFactory = _safeProxyFactory;
+        recoveryModule = _recoveryModule;
+    }
+
+    /// @dev Creates a Safe multisig.
+    /// @param owners Set of multisig owners.
+    /// @param threshold Number of required confirmations for a multisig transaction.
+    /// @param data Encoded data related to the creation of a chosen multisig.
+    /// @return multisig Address of a created multisig.
+    function create(
```

Comment:
any reason we don't do a symmetric "parseData" implementation to this https://github.com/valory-xyz/autonolas-registries/blob/main/contracts/multisigs/GnosisSafeMultisig.sol#L45 ? would make comparisons trivial

Reply:
To be honest there's no need for that. We can only parse `fallbackHandler` address and `nonce`. We don't want to parse any additional payload since we only allow the module related one. All the rest of params - those are useless as gas tokens do not exist for a long time already, and nobody is going to send any tokens during the Safe setup.

Reply:
FYI re: gas tokens, thanks @77ph for the link:
https://blog.openzeppelin.com/fundamentals-of-gas-tokens

I believe that part of the code is left even in latest version of Safe implementation for legacy purposes only.

Reply:
ok